File: test/ipfs_accelerate_py/api_backends/sample_backend.py
import os
import requests
import json
import logging
from typing import Dict, List, Any, Optional, Union

# Setup logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("sample_backend")

class sample_backend:
    """Sample API backend implementation for testing"""
    
    def __init__(self, resources: Dict[str, Any] = None, metadata: Dict[str, Any] = None):
        """Initialize the API backend"""
        self.resources = resources or {}
        self.metadata = metadata or {}
        self.api_endpoint = "https://api.sample.com/v1/chat"
        self.default_model = "sample-model"
        self.queue = []
        self.circuit_open = False
        self.error_count = 0
        
    def get_api_key(self, metadata: Dict[str, Any]) -> str:
        """Get API key from metadata or environment"""
        return metadata.get("sample_api_key") or os.environ.get("SAMPLE_API_KEY", "")
    
    def get_default_model(self) -> str:
        """Get the default model name"""
        return self.default_model
    
    def is_compatible_model(self, model: str) -> bool:
        """Check if the model is compatible with this backend"""
        return model.startswith("sample")
    
    def create_endpoint_handler(self) -> Callable:
        """Create an endpoint handler function"""
        def handler(data: Dict[str, Any]) -> Dict[str, Any]:
            return self.make_post_request(data)
        return handler
    
    def test_endpoint(self) -> bool:
        """Test the API endpoint"""
        try:
            api_key = self.get_api_key(self.metadata)
            if not api_key:
                logger.error("API key is required")
                return False
                
            test_request = {
                "model": self.get_default_model(),
                "messages": [{"role": "user", "content": "Hello"}],
                "max_tokens": 5
            }
            
            response = self.make_post_request(test_request, api_key)
            return True
        except Exception as e:
            logger.error(f"Endpoint test failed: {e}")
            return False
    
    def make_post_request(self, data: Dict[str, Any], api_key: Optional[str] = None, 
                          timeout: int = 30) -> Dict[str, Any]:
        """Make a POST request to the API"""
        api_key = api_key or self.get_api_key(self.metadata)
        if not api_key:
            raise ValueError("API key is required")
            
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {api_key}"
        }
        
        response = requests.post(
            self.api_endpoint,
            headers=headers,
            json=data,
            timeout=timeout
        )
        
        if response.status_code != 200:
            error_data = response.json() if response.content else {}
            raise Exception(f"API request failed: {error_data}")
            
        return response.json()
    
    def make_stream_request(self, data: Dict[str, Any], api_key: Optional[str] = None,
                           timeout: int = 30) -> Iterator[Dict[str, Any]]:
        """Make a streaming request to the API"""
        api_key = api_key or self.get_api_key(self.metadata)
        if not api_key:
            raise ValueError("API key is required")
            
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {api_key}",
            "Accept": "text/event-stream"
        }
        
        # Ensure stream parameter is set
        data["stream"] = True
        
        response = requests.post(
            self.api_endpoint,
            headers=headers,
            json=data,
            timeout=timeout,
            stream=True
        )
        
        if response.status_code != 200:
            error_data = response.json() if response.content else {}
            raise Exception(f"API stream request failed: {error_data}")
            
        for line in response.iter_lines():
            if line:
                line_text = line.decode('utf-8')
                if line_text.startswith('data: '):
                    data = line_text[6:]
                    if data == '[DONE]':
                        break
                    try:
                        chunk = json.loads(data)
                        yield chunk
                    except json.JSONDecodeError:
                        logger.error(f"Failed to parse stream data: {data}")
    
    def chat(self, messages: List[Dict[str, str]], model: Optional[str] = None,
             max_tokens: Optional[int] = None, temperature: Optional[float] = None,
             top_p: Optional[float] = None) -> Dict[str, Any]:
        """Generate a response using the API"""
        model = model or self.get_default_model()
        
        request_data = {
            "model": model,
            "messages": messages,
            "max_tokens": max_tokens,
            "temperature": temperature,
            "top_p": top_p
        }
        
        # Remove None values from request data
        request_data = {k: v for k, v in request_data.items() if v is not None}
        
        response = self.make_post_request(request_data)
        
        return {
            "id": response.get("id", ""),
            "model": response.get("model", model),
            "object": response.get("object", "chat.completion"),
            "created": response.get("created", 0),
            "choices": response.get("choices", []),
            "usage": response.get("usage", {})
        }
    
    def stream_chat(self, messages: List[Dict[str, str]], model: Optional[str] = None,
                    max_tokens: Optional[int] = None, temperature: Optional[float] = None,
                    top_p: Optional[float] = None) -> Iterator[Dict[str, Any]]:
        """Generate a streaming response using the API"""
        model = model or self.get_default_model()
        
        request_data = {
            "model": model,
            "messages": messages,
            "max_tokens": max_tokens,
            "temperature": temperature,
            "top_p": top_p,
            "stream": True
        }
        
        # Remove None values from request data
        request_data = {k: v for k, v in request_data.items() if v is not None}
        
        for chunk in self.make_stream_request(request_data):
            yield chunk

File: test/tests/web/ipfs_accelerate_js_tests/src/api_backends/sample_backend/sample_backend.ts
import { BaseApiBackend } from '../base';
import { ApiMetadata, ApiRequestOptions, Message, ChatCompletionResponse, StreamChunk } from '../types';
import { SampleBackendResponse, SampleBackendRequest } from './types';

export class SampleBackend extends BaseApiBackend {

  private apiEndpoint: string = 'https://api.samplebackend.com/v1/chat';

  constructor(resources: Record<string, any> = {}, metadata: ApiMetadata = {}) {
    super(resources, metadata);
  }

protected getApiKey(metadata: ApiMetadata): string {
  return metadata.sample_backend_api_key || 
         metadata.sampleBackendApiKey || 
         (typeof process !== 'undefined' ? process.env.SAMPLE_BACKEND_API_KEY || '' : '');
}
protected getDefaultModel(): string {
  return "sample_backend-model";
}
isCompatibleModel(model: string): boolean {
  // Return true if this model is compatible with this backend
  return model.toLowerCase().includes('sample_backend');
}
  createEndpointHandler(): (data: any) => Promise<any> {
    return async (data: any) => {
      try {
        return await this.makePostRequest(data);
      } catch (error) {
        throw this.createApiError(`${this.constructor.name} endpoint error: ${error.message}`, 500);
      }
    };
  }

  async testEndpoint(): Promise<boolean> {
    try {
      const apiKey = this.getApiKey(this.metadata);
      if (!apiKey) {
        throw this.createApiError('API key is required', 401, 'authentication_error');
      }

      // Make a minimal request to verify the endpoint works
      const testRequest = {
        model: this.getDefaultModel(),
        messages: [{ role: 'user', content: 'Hello' }],
        max_tokens: 5
      };

      await this.makePostRequest(testRequest, apiKey, { timeoutMs: 5000 });
      return true;
    } catch (error) {
      console.error(`${this.constructor.name} endpoint test failed:`, error);
      return false;
    }
  }

  async makePostRequest(data: Record<string, any><[string, any]>, api_key: string | null | undefined = 30, timeout: number = 30): Promise<any> {
    const apiKey = apiKey || this.getApiKey(this.metadata);
    if (!apiKey) {
      throw this.createApiError('API key is required', 401, 'authentication_error');
    }

    // Process with queue and circuit breaker
    return this.processWithQueueAndBackoff(async () => {
      // Prepare request headers
      const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${apiKey}`
      };

      // Prepare request body
      const requestBody = JSON.stringify(data);

      // Set up timeout
      const timeoutMs = options?.timeoutMs || 30000;
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), timeoutMs);

      try {
        // Make the request
        const response = await fetch(this.apiEndpoint, {
          method: 'POST',
          headers,
          body: requestBody,
          signal: controller.signal
        });

        // Check for errors
        if (!response.ok) {
          const errorData = await response.json().catch(() => ({}));
          throw this.createApiError(
            errorData.error?.message || `HTTP error ${response.status}`,
            response.status,
            errorData.error?.type || 'api_error'
          );
        }

        // Parse response
        const responseData = await response.json();
        return responseData;
      } catch (error) {
        if (error.name === 'AbortError') {
          throw this.createApiError(`Request timed out after ${timeoutMs}ms`, 408, 'timeout_error');
        }
        throw error;
      } finally {
        clearTimeout(timeoutId);
      }
    }, options);
  }

  async *makeStreamRequest(data: Record<string, any><[string, any]>, api_key: string | null | undefined = 30, timeout: number = 30): AsyncGenerator<StreamChunk> {
    const apiKey = this.getApiKey(this.metadata);
    if (!apiKey) {
      throw this.createApiError('API key is required', 401, 'authentication_error');
    }

    // Ensure stream option is set
    const streamData = { ...data, stream: true };

    // Prepare request headers
    const headers = {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${apiKey}`
    };

    // Prepare request body
    const requestBody = JSON.stringify(streamData);

    // Set up timeout
    const timeoutMs = options?.timeoutMs || 30000;
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeoutMs);

    try {
      // Make the request
      const response = await fetch(this.apiEndpoint, {
        method: 'POST',
        headers,
        body: requestBody,
        signal: controller.signal
      });

      // Check for errors
      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw this.createApiError(
          errorData.error?.message || `HTTP error ${response.status}`,
          response.status,
          errorData.error?.type || 'api_error'
        );
      }

      if (!response.body) {
        throw this.createApiError('Response body is null', 500, 'stream_error');
      }

      // Process the stream
      const reader = response.body.getReader();
      const decoder = new TextDecoder();
      let buffer = '';

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split('\n');
        buffer = lines.pop() || ''; // Keep the last incomplete line in the buffer

        for (const line of lines) {
          if (line.trim() === '') continue;
          if (line.startsWith('data: ')) {
            const data = line.slice(6);
            if (data === '[DONE]') continue;

            try {
              const parsed = JSON.parse(data);
              yield {
                content: parsed.choices?.[0]?.delta?.content || '',
                type: 'delta'
              };
            } catch (e) {
              console.warn('Failed to parse stream data:', data);
            }
          }
        }
      }

      // Handle any remaining data in the buffer
      if (buffer.trim() !== '') {
        if (buffer.startsWith('data: ') && buffer !== 'data: [DONE]') {
          try {
            const data = buffer.slice(6);
            const parsed = JSON.parse(data);
            yield {
              content: parsed.choices?.[0]?.delta?.content || '',
              type: 'delta'
            };
          } catch (e) {
            console.warn('Failed to parse final stream data:', buffer);
          }
        }
      }
    } catch (error) {
      if (error.name === 'AbortError') {
        throw this.createApiError(`Stream request timed out after ${timeoutMs}ms`, 408, 'timeout_error');
      }
      throw error;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  async chat(messages: Array<any><Record<string, any><[string, string]>>, model: string | null | undefined = null, max_tokens: number | null | undefined = null, temperature: number | null | undefined = null, top_p: number | null | undefined = null): Promise<ChatCompletionResponse> {
    // Prepare request data
    const modelName = options?.model || this.getDefaultModel();
    
    const requestData: SampleBackendRequest = {
      model: modelName,
      messages,
      max_tokens: options?.maxTokens,
      temperature: options?.temperature,
      top_p: options?.topP
    };

    // Make the request
    const response = await this.makePostRequest(requestData, undefined, options);

    // Convert to standard format
    return {
      id: response.id || '',
      model: response.model || model,
      content: response.choices?.[0]?.message?.content || '',
      created: response.created || Date.now(),
      usage: response.usage || { prompt_tokens: 0, completion_tokens: 0, total_tokens: 0 }
    };
  }

  async *streamChat(messages: Array<any><Record<string, any><[string, string]>>, model: string | null | undefined = null, max_tokens: number | null | undefined = null, temperature: number | null | undefined = null, top_p: number | null | undefined = null): AsyncGenerator<StreamChunk> {
    // Prepare request data
    const modelName = options?.model || this.getDefaultModel();
    
    const requestData: SampleBackendRequest = {
      model: modelName,
      messages,
      stream: true,
      max_tokens: options?.maxTokens,
      temperature: options?.temperature,
      top_p: options?.topP
    };

    // Make the stream request
    const stream = this.makeStreamRequest(requestData, options);
    
    // Pass through the stream chunks
    return stream;
  }

}